galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -233,41 +233,74 @@ def get(self, request, *args, **kwargs):
             return redirect(distribution.content_guard.cast().preauthenticate_url(url))
 
 
-class CollectionVersionMoveViewSet(api_base.ViewSet):
-    permission_classes = [access_policy.CollectionAccessPolicy]
-
-    def move_content(self, request, *args, **kwargs):
-        """Remove content from source repo and add to destination repo.
+class CollectionRepositoryMixing:
 
-        Creates new RepositoryVersion of source repo without content included.
-        Creates new RepositoryVersion of destination repo with content included.
-        """
+    @property
+    def version_str(self):
+        """Build version_str from request."""
+        return '-'.join([self.kwargs[key] for key in ('namespace', 'name', 'version')])
 
-        version_str = '-'.join([self.kwargs[key] for key in ('namespace', 'name', 'version')])
+    def get_collection_version(self):
+        """Get collection version entity."""
         try:
-            collection_version = CollectionVersion.objects.get(
+            return CollectionVersion.objects.get(
                 namespace=self.kwargs['namespace'],
                 name=self.kwargs['name'],
                 version=self.kwargs['version'],
             )
         except ObjectDoesNotExist:
-            raise NotFound(_('Collection %s not found') % version_str)
+            raise NotFound(_('Collection %s not found') % self.version_str)
 
+    def get_repos(self):
+        """Get src and dest repos."""
         try:
             src_repo = AnsibleDistribution.objects.get(
                 base_path=self.kwargs['source_path']).repository
             dest_repo = AnsibleDistribution.objects.get(
                 base_path=self.kwargs['dest_path']).repository
         except ObjectDoesNotExist:
-            raise NotFound(_('Repo(s) for moving collection %s not found') % version_str)
+            raise NotFound(_('Repo(s) for moving collection %s not found') % self.version_str)
+        return src_repo, dest_repo
 
-        content_obj = Content.objects.get(pk=collection_version.pk)
 
+class CollectionVersionCopyViewSet(api_base.ViewSet, CollectionRepositoryMixing):
+    permission_classes = [access_policy.CollectionAccessPolicy]
+
+    def copy_content(self, request, *args, **kwargs):
+        """Copy collection version from one  repository to another."""
+
+        collection_version = self.get_collection_version()
+        src_repo, dest_repo = self.get_repos()
+        copy_task = dispatch(
+            add_and_remove,
+            exclusive_resources=[src_repo, dest_repo],
```

Comment:
I'm not sure `src_repo` needs exclusive access, could probably be moved to `shared_resources`

https://docs.pulpproject.org/pulpcore/en/latest/plugins/api-reference/tasking.html#pulpcore.plugin.tasking.dispatch

Reply:
done